openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb

```diff
@@ -0,0 +1,509 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79021778",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_net.to('hpu')\n",
+    "params_to_update = []\n",
+    "for param in model_net.parameters():\n",
+    "    if param.requires_grad == True:\n",
+    "        params_to_update.append(param)\n",
+    "        \n",
+    "optimizer_adam = optim.Adam(params_to_update, lr=1e-4)\n",
+    "\n",
+    "def cross_entropy(output, target):\n",
+    "    \"\"\"Binary cross-entropy metric\n",
+    "    \"\"\"\n",
+    "    return F.cross_entropy(input=output,target=target)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f097cdc5",
+   "metadata": {},
+   "source": [
+    "### Register model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "06a8cca8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "framework_adapter = 'openfl.plugins.frameworks_adapters.pytorch_adapter.FrameworkAdapterPlugin'\n",
+    "model_interface = ModelInterface(model=model_net, optimizer=optimizer_adam, framework_plugin=framework_adapter)\n",
+    "\n",
+    "# Save the initial model state\n",
+    "initial_model = deepcopy(model_net)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "849c165b",
+   "metadata": {},
+   "source": [
+    "## Define and register FL tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b9649385",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "task_interface = TaskInterface()\n",
+    "\n",
+    "\n",
+    "# The Interactive API supports registering functions definied in main module or imported.\n",
+    "def function_defined_in_notebook(some_parameter):\n",
+    "    print(f'Also I accept a parameter and it is {some_parameter}')\n",
+    "\n",
+    "# Task interface currently supports only standalone functions.\n",
+    "@task_interface.add_kwargs(**{'some_parameter': 42, 'device':'hpu'})\n",
+    "@task_interface.register_fl_task(model='net_model', data_loader='train_loader', \\\n",
+    "                     device='device', optimizer='optimizer')     \n",
+    "def train(net_model, train_loader, optimizer, device, loss_fn=cross_entropy, some_parameter=None):\n",
+    "    torch.manual_seed(0)\n",
+    "    function_defined_in_notebook(some_parameter)\n",
+    "    \n",
+    "    train_loader = tqdm.tqdm(train_loader, desc=\"train\")\n",
+    "    net_model.train()\n",
+    "    net_model.to(device)\n",
+    "\n",
+    "    losses = []\n",
+    "\n",
+    "    for data, target in train_loader:\n",
+    "        data, target = torch.tensor(data).to(device), torch.tensor(\n",
+    "            target).to(device)\n",
+    "        optimizer.zero_grad()\n",
+    "        output = net_model(data)\n",
+    "        loss = loss_fn(output=output, target=target)\n",
+    "        loss.backward()\n",
+    "        # API call to trigger execution\n",
+    "        htcore.mark_step()\n",
```

Comment:
Same as previous comment for `NOTE`

Reply:
This is not a workaround, it is part of the HPU adaptations steps as mentioned in README.md